### ISSUE TITLE
Connection to Redis not closed properly

### ISSUE THREAD

**Issue by stevedomin** [getsentry/sentry#2826]
Sentry: 8.1.1
Redis: v=3.0.7 sha=00000000:0 malloc=jemalloc-3.6.0 bits=64 build=118196f85002221b

Since we have upgraded to 8.1.1 we see a lot of socket connections to Redis in the CLOSE_WAIT state.

Here's what a `netstat -ntp | grep CLOSE_WAIT` looks like (we currently have 600 or so of these lines):

```
[...]
tcp        1      0 x.x.x.x:45789      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:54185      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:47213      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
tcp        1      0 x.x.x.x:33223      x.x.x.x:6379       CLOSE_WAIT  19996/[Sentry] uWSG
tcp        1      0 x.x.x.x:36270      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:33117      x.x.x.x:6379       CLOSE_WAIT  19995/[Sentry] uWSG
tcp        1      0 x.x.x.x:38550      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
tcp        1      0 x.x.x.x:36268      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
tcp        1      0 x.x.x.x:56623      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
tcp        1      0 x.x.x.x:57586      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:49732      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:48513      x.x.x.x:6379       CLOSE_WAIT  20018/[celeryd: cel
tcp        1      0 x.x.x.x:55005      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
tcp        1      0 x.x.x.x:56749      x.x.x.x:6379       CLOSE_WAIT  19996/[Sentry] uWSG
tcp        1      0 x.x.x.x:41307      x.x.x.x:6379       CLOSE_WAIT  20017/[celeryd: cel
[...]
```

**Comment by mattrobenolt** [comment]
Just for a point of reference, what version were you running before?

**Comment by stevedomin** [comment]
@mattrobenolt 7.7.4

**Comment by mattrobenolt** [comment]
Oh boy. :) Does your `CLOSE_WAIT' correspond with `CLIENT LIST` from Redis? http://redis.io/commands/client-list

**Comment by stevedomin** [comment]
I've upgraded to 8.2.2 and it seems to have fixed the issue.

I haven't checked the `CLIENT LIST` from Redis when we are having these issues so I couldn't tell you unfortunately.

**Comment by dcramer** [comment]
Feel free to reopen if you do see this happening again.

**Comment by stevedomin** [comment]
Will do, thanks @dcramer

**Comment by stevedomin** [comment]
Should have checked before replying. The list is growing slower but it's still growing. 157 connections in `CLOSE_WAIT` at the moment.

I really have no idea what's causing this.

**Comment by mattrobenolt** [comment]
Can you check now if this correlates to `CLIENT LIST`? :)

**Comment by stevedomin** [comment]
Just checked, they don't match, I've 158 connections in `CLOSE_WAIT` for 14 Redis clients

**Comment by mattrobenolt** [comment]
That's interesting then. We don't see this at all on any of our redis servers in production for getsentry.com.

/cc @mitsuhiko because he may have an idea. The biggest change from 7.7 to 8.x in this area is the introduction of https://github.com/getsentry/rb for redis connection management.

**Comment by nlopes** [comment]
Any idea of what's happening in this ticket? 
```
tcp        1      0 X:54957      X:6379       CLOSE_WAIT  1665/[Sentry] uWSGI
tcp        1      0 X:54871      X:6379       CLOSE_WAIT  1829/sentry.conf.py
tcp        1      0 X:54872      X:6379       CLOSE_WAIT  1829/sentry.conf.py
tcp        1      0 X:54858      X:6379       CLOSE_WAIT  1564/[Sentry] uWSGI
```

Within half a day, these go up to ~30 sockets in CLOSE_WAIT. If you need additional details to help troubleshoot, let me know, I'll gladly get the info needed.

Version of `rb` from `pip freeze` is 1.4

**Comment by nlopes** [comment]
Also, and to give an idea of how bad this is (red is connections in CLOSE_WAIT and where you see the change is when we moved to sentry 8):

![screen shot 2016-04-06 at 14 46 36](https://cloud.githubusercontent.com/assets/652790/14318813/77618d28-fc06-11e5-8258-c2cf56298e99.png)

**Comment by nlopes** [comment]
Quick update here:

I've decided to upgrade kombu to 3.0.35 and celery to 3.1.23 even though those versions fall outside the versions set in setup.py.

What prompted me to do that was this comment: https://github.com/celery/celery/issues/1845#issuecomment-174052659

I may have introduced other issues by bumping those versions and I'd appreciate some feedback on that from @mattrobenolt or @mitsuhiko.

This is what the last day looks like (11:20 was when I bumped kombu and celery versions manually):
![screen shot 2016-04-07 at 09 28 38](https://cloud.githubusercontent.com/assets/652790/14345127/445c40d8-fca3-11e5-9aeb-924109fdb279.png)

We still have around 15-22 sockets in CLOSE-WAIT state at any given point in time but at least they no longer seem to be unbounded.
